var/spack/repos/builtin/packages/gromacs/package.py

```diff
@@ -276,7 +284,11 @@ class Gromacs(CMakePackage, CudaPackage):
     depends_on("pkgconfig", type="build")
 
     depends_on("cuda", when="+cuda")
-    depends_on("sycl", when="+sycl")
+    with when("+rocm"):
```

Comment:
While having `+rocm` variant very nice for ease of use and consistency with other packages, it is, in my opinion, misleading to have it as a "first class citizen"; rather, it should be an alias to something like `+sycl+acpp+rocm`.

There are other ways to build GROMACS with ROCm:

- The recommended way (for GROMACS `@2022:2024`) of using SYCL/ACpp/ROCm, implemented here.
- SYCL/oneAPI/ROCm, less performant at the moment, but supported.
- There is work underway of adding native HIP backend in GROMACS, which, in some form, will be available in GROMACS 2025.
- AMD's OpenCL is a part of ROCm stack, so `+opencl` also, technically, falls under `+rocm` (and is the only supported way for GROMACS `@:2021`).

Renaming this flag to `acpp` and making `rocm` an "alias" (not sure how to best call it in Spack terms) would already clear a lot of the confusion and require only minimal changes to the logic here.

Reply:
I agree but I am not sure how to express that in Spack. @haampie @tgamblin any ideas?

Reply:
> it should be an alias to something like `+sycl+acpp+rocm`

I'm not clear what this means: I would take `+sycl+rocm` to be "sycl-with-ROCm" and `+sycl+cuda` to be "sycl-with-CUDA", do you have examples of expressions that are unreasonable with a first-class `+rocm` variant?

> There are other ways to build GROMACS with ROCm:

I'm not sure how the `+rocm` variant removes the possibility of supporting these: there could be an additional `rocm_backend` variant like:

```
with when("+rocm"):
    variant("rocm_backend", default="hip", values=("hip", "opencl", "oneapi", "native-hip"))
```

Reply:
> I'm not clear what this means: I would take +sycl+rocm to be "sycl-with-ROCm" and +sycl+cuda to be "sycl-with-CUDA", do you have examples of expressions that are unreasonable with a first-class +rocm variant?

I was concerned whether Spack can handle such cases fine. But if one can use either `+rocm` or `+sycl+rocm` or `+sycl+acpp+rocm` to mean one thing, and `+sycl+oneapi+rocm` to mean another, then there should be no problem.

> I'm not sure how the +rocm variant removes the possibility of supporting these: there could be an additional rocm_backend variant like:

It does not remove any possibility, but it, IMO, makes the build configuration confusing and not quite representable of how different runtimes are layered.

oneAPI is not a ROCm backend; ROCm is one of oneAPI backends. Or one of AdaptiveCpp's backends. At least as far as GROMACS is concerned.

However, there is, for example, chipStar/HIPLZ project that aims to enable running ROCm applications on Intel GPUs. In this case, oneAPI *is* a backend of ROCm.

And while it is not ambiguous for GROMACS (no plans to support chipStar whatsoever), I think it is worthwhile to try to find a generic solution that can later be adopted by other projects.

And on topic of generic. We are exploring the use of AdaptiveCpp's SSCP backend, which builds generic binaries that can JIT to any GPU target (and, thanks to JIT optimizations, occasionally outperforms AOT compilation, at least in benchmarks). In this case, we would not need to know `amdgpu_target` but `ROCmPackage` would still not tolerate having it `none`, if I understand correctly (one can supply an arbitrary value to bypass the check, but that sounds like a kludge).

Reply:
> if one can use either +rocm or +sycl+rocm or +sycl+acpp+rocm to mean one thing, and +sycl+oneapi+rocm to mean another, then there should be no problem.

Yes, in the sense that all phase methods can access all of these variant values (so when generating CMake options, one can add whatever `if` statements they want) - sorry if I missed a finer point you were making here.

> It does not remove any possibility, but it, IMO, makes the build configuration confusing and not quite representable of how different runtimes are layered.

Ah, I think I chose the wrong name then: it could be something like

```
with when("+rocm"):
    variant("rocm_method", default="hipsycl", values=("hipsycl", "oneapi", "native-hip"))
```

so for a couple of these ROCm is a backend and in at least one case (for upcoming GROMACS 2025) it would be direct.